### PR TITLE
Webhook Signature Verification

### LINKED ISSUE #83: Support for Signed Web Hooks?
Stripe supports [verifying Web Hook payloads via HMAC SHA-256 signatures](https://stripe.com/docs/webhooks#signatures). The `stripe-ruby` gem recently (on 2017-04-28) added support for verifying signatures. Now that it's available, perhaps it can be rolled into `stipe_event`? For example, perhaps the default retriever can first verify the signature before trying retrieve the event from Stripe?

Before I put any effort toward a PR to add this, I wanted to make sure it was something you were open to.

Thank you.

### PR DESCRIPTION
**Pull request by mikeycgto**
Closes #83.

This to do for this PR:

- [x] Update the README
- [x] Get my specs running and add a few for this feature

### CODE REVIEW

**Comment by rmm5t** [comment]
> - Add deprecation errors and/or warnings for removal of `StripeEvent.authentication_secret`

Let's keep things focused. Please don't remove or or change the behavior of the `authentication_secret` as part of this pull-request.  Let's only consider deprecating it after we get signature verification fully working and released. The deprecation deserves to be part of a separate PR later.

**Comment by mikeycgto** [comment]
Sounds good. Had only included it in this PR cause the original issue mentioned that `authentication_secret` can be deprecated. I will remove that portion of the PR and keep both features in place.

Need to get the specs running locally so I can add some for this feature. Was having bundler version issues when I tried to run them initially.

**Comment by rmm5t** [comment]
Thanks for making the adjustments to this PR. Looks clean. Have you tried this branch against a live Stripe endpoint yet?  

My only critique is that I'd prefer the tests to pass an actual HMAC signature header instead of mocking the call to `Stripe::Webhook::Signature.verify_header` (I prefer a classical style of testing where possible vs a mockist style), but there's no need for that to hold up this PR.

I can merge and release a new version later today if no one else objects. 

Good work @mikeycgto!

**Comment by mikeycgto** [comment]
Thanks!

Yeah I do agree with you and prefer to avoid mocking when possible. My concern there was the specs would become to specific to actual the HMAC implementation in Stripe. Figured mocks would just be quicker to write and more future-proof if their verification process changes (assuming the API doesn't change :smiley:).

**Comment by brandur** [comment]
@rmm5t @mikeycgto Nice! I'd +1 trying to go a level deeper than the mock if possible. What you can do in this case is to call `Stripe::Webhook::Signature.compute_signature(payload, secret)` to generate a valid signature before validating it. That way you can test more deeply than the mock and you don't have to rely on any particular crypto implementation — if we update to something beyond HMAC-SHA256, we'll be updating `.compute_signature` as well `.verify_header`.

**Comment by mikeycgto** [comment]
@brandur this was more involved than I thought it would be (have to include a timestamp and version scheme). I do agree it's better to use real signatures here though. Thanks for the feedback!

File: spec/controllers/webhook_controller_spec.rb [hunk 1]
```diff
@@ -84,4 +84,38 @@ def webhook_with_secret(secret, params)
...
+
+    let(:secret) { 'secret' }
+    let(:signature) { 'signature' }
+    let(:timestamp) { Time.now.to_i }
+
+    before(:each) { StripeEvent.signing_secret = secret }
+    after(:each) { StripeEvent.signing_secret = nil }
+
+    it "rejects invalid signatures" do
+      webhook_with_signature generate_signature('invalid payload'), id: 'evt_charge_succeeded'
```
Comment (brandur) [inline spec/controllers/webhook_controller_spec.rb]: Hm, I don't think I fully understand this test. Why does this one come out as invalid with the one below valid? It seems that it should be computing a valid signature for any value that you send to `#generate_signature` shouldn't it?

File: spec/controllers/webhook_controller_spec.rb [hunk 2]
```diff
@@ -84,4 +84,38 @@ def webhook_with_secret(secret, params)
...
+    end
+
+    def generate_signature(payload)
+      signature = Stripe::Webhook::Signature.send(:compute_signature, "#{timestamp}.#{payload}", secret)
+
+      "t=#{timestamp},v1=#{signature}"
+    end
+
+    let(:secret) { 'secret' }
+    let(:signature) { 'signature' }
```
Comment (brandur) [inline spec/controllers/webhook_controller_spec.rb]: I think this might be unused now.

File: spec/controllers/webhook_controller_spec.rb [hunk 3]
```diff
@@ -84,4 +84,38 @@ def webhook_with_secret(secret, params)
...
   end
+
+  context "with a signing secret" do
+    def webhook_with_signature(signature, params)
+      request.env['HTTP_STRIPE_SIGNATURE'] = signature
+      webhook params
+    end
+
+    def generate_signature(payload)
+      signature = Stripe::Webhook::Signature.send(:compute_signature, "#{timestamp}.#{payload}", secret)
```
Comment (brandur) [inline spec/controllers/webhook_controller_spec.rb]: Shoot. I forgot this is declared private. Sorry about that! I expect this to be pretty stable, but we might want to consider exposing this publicly for testing purposes if nothing else.

File: spec/controllers/webhook_controller_spec.rb (hunk 2 above)
Comment (mikeycgto) [reply spec/controllers/webhook_controller_spec.rb]: Indeed 😄

File: spec/controllers/webhook_controller_spec.rb (hunk 1 above)
Comment (mikeycgto) [reply spec/controllers/webhook_controller_spec.rb]: Perhaps I shouldn't use `generate_signature`? I wanted to have a spec that passes the correct format with the timestamp but with an incorrect signature. Maybe I should just pass a string directly here like so:

```ruby
webhook_with_signature "t=#{timestamp},v1=invalid", id: 'evt_charge_succeeded'
```

Perhaps a better spec would be to `"reject missing signature"` where no header is provided? Any suggestions here @rmm5t?

**Comment by mikeycgto** [comment]
Refactored the specs some to hopefully make more sense. Thanks again for the feedback!

File: spec/controllers/webhook_controller_spec.rb (hunk 1 above)
Comment (rmm5t) [reply spec/controllers/webhook_controller_spec.rb]: I would suggest just passing in a "bogus" signature string and avoid the call to `generate_signature`.

**Comment by rmm5t** [comment]
@mikeycgto Looks good to me. Are you done with your refactoring? I'm happy to merge now when ready.

**Comment by mikeycgto** [comment]
I'm good with merging. Thanks.

**Pull request merged**